### PR TITLE
Remove unnecessary inner interface and improve javadoc on local plugin resolution types

### PR DESCRIPTION
**Pull request by tresat**
The inner `Reference` type and its implementation seem to be unnecessary.  This PR simplifies the `ProjectPublicationRegistry` slightly by removing them.

I added tests to demonstrate the behavior when multiple plugin builds are included reusing the same project name within the build tree.  This does not appear to be a problem.

### CODE REVIEW

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

File: subprojects/composite-builds/src/integTest/groovy/org/gradle/integtests/composite/plugins/PluginBuildsIntegrationTest.groovy [hunk 1]
```diff
@@ -624,4 +625,114 @@ class PluginBuildsIntegrationTest extends AbstractPluginBuildIntegrationTest {
         where:
         dsl << ['Groovy', 'Kotlin']
     }
+
+    def "multiple included plugin builds which each include a library build with the same name works with #dsl dsl"() {
```
Comment (big-guy) [inline subprojects/composite-builds/src/integTest/groovy/org/gradle/integtests/composite/plugins/PluginBuildsIntegrationTest.groovy:629]: I'm not sure if this is testing anything in particular.  The `lib`s are identical, so if we were getting the "wrong" one, you wouldn't notice.

File: subprojects/composite-builds/src/integTest/groovy/org/gradle/integtests/composite/plugins/AbstractPluginBuildIntegrationTest.groovy [hunk 2]
```diff
@@ -41,6 +45,11 @@ abstract class AbstractPluginBuildIntegrationTest extends AbstractIntegrationSpe
         final TestFile projectPluginFile
 
         PluginBuildFixture(String buildName, boolean useKotlinDSL) {
+            this(null, buildName, useKotlinDSL)
+        }
+
+        PluginBuildFixture(String rootDir, String buildName, boolean useKotlinDSL) {
```
Comment (big-guy) [inline subprojects/composite-builds/src/integTest/groovy/org/gradle/integtests/composite/plugins/AbstractPluginBuildIntegrationTest.groovy:51]: I don't think this is a good idea.  We have other fixtures that do something like this and the String concatenation makes it hard to follow what paths are being used. 

Maybe something like `TestFile rootDir` could be used here. But we'd need to make other changes:
1. PluginBuildFixture would be made static
2. The default for rootDir would be the temporary test directory from the outer test
3. all calls to `file(...)` below would need to be `rootDir.file(...)`. This avoids the string concatenation problem.
4. There might be other changes that I haven't noticed

File: subprojects/composite-builds/src/integTest/groovy/org/gradle/integtests/composite/plugins/PluginBuildsIntegrationTest.groovy [hunk 3]
```diff
@@ -624,4 +625,114 @@ class PluginBuildsIntegrationTest extends AbstractPluginBuildIntegrationTest {
...
+        pluginBuild1.settingsFile.text = """
+            pluginManagement {
+                includeBuild("${libBuild1.buildName}")
+            }
+        """
+        pluginBuild1.buildFile << """
+            plugins {
+                id("${libBuild1.projectPluginId}")
+            }
+        """
```
Comment (big-guy) [inline subprojects/composite-builds/src/integTest/groovy/org/gradle/integtests/composite/plugins/PluginBuildsIntegrationTest.groovy:699]: isn't this what the test fixture should already be doing?

oh, I think I see... libBuild1 has a plugin, but it's nested deeper.  pluginBuild1 has a plugin.

You're adding libBuild1 as another included build, which should have the same plugin id. I think this is a case where the test fixtures are doing us no favors.

**Review by big-guy — CHANGES_REQUESTED** [review]
I think we should abandon this. It looks like a time sink to figure out how to make this better for local plugin resolution.

**Comment by tresat** [comment]
> I think we should abandon this. It looks like a time sink to figure out how to make this better for local plugin resolution.

OK. Abandoning.
